src/iterative_wrappers.jl

```diff
@@ -226,6 +226,11 @@ end
 
 function SciMLBase.solve!(cache::LinearCache, alg::KrylovJL; kwargs...)
     if cache.isfresh
+        if hasproperty(alg, :precs) && !isnothing(alg.precs)
+            Pl, Pr = cache.alg.precs(x, cache.p)
+            cache.Pl = Pl
+            cache.Pr = Pr
+        end
```

Comment:
And generally, could we have it like this:
```
function SciMLBase.solve!(cache::LinearCache, alg::KrylovJL; reuse_precs=true, kwargs...)
    if cache.isfresh
        if hasproperty(alg, :precs) && !isnothing(alg.precs)
            if reuse_precs || cache.Pl == nothing || cache.Pr == nothing
                Pl, Pr = cache.alg.precs(cache.A, cache.p)
                cache.Pl = Pl
                cache.Pr = Pr
            end
        end
 ...
```

Reply:
Also, in the moment this leads to `precs` being called twice: once in `init` and once here.

Reply:
init should set fresh to false.

Reply:
ah that's tricky. I think I will need to add an extra flag to fix this.

Reply:
@ChrisRackauckas really? I think the code relies on it being true to set up some of the C libraries.

Reply:
I guess that possibly could move to init?

Reply:
So this is annoyingly complicated. The basic issue is that the LU-based factorizations put "fake" factorizations in at init time (i.e. factorizations of 1x1 matrices) presumably to reduce the time and memory cost. This means that they require `isfresh=true` after init, since during the first solve, they need to resize the arrays and do the real factorization. This sort of strategy can't work for preconditioners, since we don't know ahead of time what type the preconditioner function will return.

The only solution to this that I can think of would be to make the factorization solvers `init` do the actual factorization so when `!cache.isfresh`, all that happens is the ldiv, but this seems sub-optimal as it would make the `init` pretty expensive...

Any other options here?

Reply:
>  This sort of strategy can't work for preconditioners, since we don't know ahead of time what type the preconditioner function will return.

Make a query for that part of the API?

Reply:
the preconditioner is user generated so unless we want to require users to implement a `precondition_interface` function or something, it's somewhat hard for me to see how this would work.

Reply:
I made a PR to this branch: https://github.com/oscardssmith/LinearSolve.jl/pull/3 which shifts the problem to the question why zeroinit is needed.

Reply:
I think we have to impose some kind of interface anyways. There can be a slow fallback but a way to speed it up and some extensions.

Reply:
Generaly I indeed think that preconditioners and factorizations could be handled in the very same way. With a DirectSolver algorithm type one could have e.g. alg=DirectSolver(UMFPACKFactorization) or alg=KrylovJL_CG(ILUPreconiditioner).   However this would be a significantly breaking change now.

Reply:
Nows the time to propose and make any significant breaking change to preconditioners.

Reply:
would it be an option to make it so `init` just does the factorization? I understand that it might lead to an extra factorization for the ODE case (since they will sometimes do a first factorization with fake values), but it does seem a lot simpler if we can just say that when you init, all the things that only depend on `A` are done, and the cache_isfresh is set to false.

Reply:
What's wrong with the current way? The current way is exactly that: `init` "just does the factorization". But that's a fallback. There's a function which by default factorizes, and then it's overloaded for specific factorizations to just build the factorization object as an optimization. It's a small optimization we then add to every method that's actually used, but it's usually not a big deal. That's exactly what I'm proposing here.

Reply:
the problem with the current way is that it makes the fallback slow. if we didn't have the "fast path", we could set cache.isfresh=false which would recover the speed of the fast path for all types, as long as the user inits on a matrix they actually want to solve.

Reply:
You could just set it to false on the slow path.

Reply:
the current approach is better by 1 factorization for users that create a LinearCache with garbage data, but those users will almost always be doing lots of factorizations, minimizing the advantage. OTOH, the current approach has overhead for users who just call solve (since the fake factorization isn't free), or users of the cache form who init on data they want to solve eventually.

Reply:
we could set it to false on the slow path, but doing so would be pretty hard since init_cacheval doesn't get the cache.

Reply:
and actually users that are initing with fake matrices can init with a 1x1 themselves, which will completely remove the penalty of my suggestion